README.md
# iNPPK
## Рекомендательная система для студентов. Анализ режимов дня, еды и сна для повышения успеваемости. IT-Олимпиада 2020-21.
## Лаборатория DS&ML. Новозыбковский Профессионально-Педагогический Колледж.

# Установка
pip install -t lib google-auth google-auth-httplib2 google-api-python-client --upgrade  
pip install google-cloud-vision  
pip install bs4  
pip install kivy, kivy-md, SQLAlchemy  
pip install psycopg2-binary  
pip install pymysql  
pip install lxml  
pip install python-telegram-bot --upgrade  
pip install ffpyplayer   
pip install pillow  


# Запуск приложения в эмуляторе
cd mobile
python main.py

# Запуск приложения в телефоне
Приложение собирается под Вашу платформу.
При необходимости учитывать статистику, Вам надо зарегистрироваться на сайте https://steelfeet.ru/ и уведомить меня по любому другому каналу: vk, tm, whatsapp и т.п. Контакты в футере сайта. Я добавлю Вам допуск к статистике и редактированию профиля, в котором надо указать id ВКонтакте и/или логин в Telegram.


changelog.txt
--------- 10Nov21
mobile: v.0.3.6 [build 0045]
<PhotoFoodRec> без калорий и предзагрузки; текст бОльшей длины, разный шрифт в рекомендациях блюд, base64 в импорте и requirements 

--------- 09Nov21
mobile: v.0.3.6 [build 0043]
правильная загрузка ActionList, режим разработчика - ввод URI, а не камера; т.к. камера уже работает стабильно

mobile: v.0.3.6 [build 0044]
перенесен <AddResult> в ui.kv

mobile: v.0.3.6 [build 0045]
.buildozer 



--------- 08Nov21
buildozer spec 0.2

mobile: v.0.3.6 [build 0042]
файлы из /ui объединены в один


--------- 26Oct21
mobile: v.0.3.5 [build 0041]
дизайн от Лизы #2

mobile: v.0.3.5 [build 0042]
https://habr.com/ru/post/426269/

buildozer spec 0.1
cd Project
git clone https://github.com/steelfeet/iNPPK.git
cd /home/kivy/Project/iNPPK/mobile
buildozer android debug
https://www.libsdl.org/ - недоступен из НППК
https://vk.com/@stulovblog-tor-as-proxy-windows
https://dist.torproject.org/torbrowser/10.5.10/
tor-win64-0.4.5.10
https://www.softpedia.com/get/Security/Security-Related/Tor.shtml - архив отличается
https://socproxy.ru/proksi-dlya-windows

Tor on Ubuntu
https://askubuntu.com/questions/6522/how-to-install-tor
https://askubuntu.com/questions/324685/how-to-route-all-internet-traffic-through-tor-the-onion-router
/etc/tor/torrc

sudo apt install libc6-dev
pip3 install --upgrade Cython

The VM comes pre-installed with for:
android.sdk = 28
android.api = 19 or android.api = 27
---------- заработало, но с ошибкой

apt install libssl-dev
buildozer android clean
buildozer android debug

Could not find a version that satisfies the requirement ast, functools

pip install picamera "buildozer"
---\
под *nix не надо?



Не запускается на телефоне
рабочая ВМ: D:\kivy-vm\kivy-vm\kivy-vm.vbox
VS Code
https://ip-calculator.ru/blog/ask/kak-ustanovit-visual-studio-code-v-linux/
sudo snap install --classic code
sudo apt-get install libjpeg-dev zlib1g-dev
sudo apt-get install libavdevice-dev
pip3 install kivymd
pip3 install opencv-python

buildozer android clean
buildozer android debug

adb install /home/kivy/Project/iNPPK/mobile/bin/iNPPK-0.3.5-armeabi-v7a-debug.apk

Failed to install PACKAGE.NAME.apk: Failure [INSTALL_FAILED_NO_MATCHING_ABIS: Failed to extract native libraries, res=-113]
https://www.linuxuprising.com/2018/07/anbox-how-to-install-google-play-store.html
https://github.com/geeks-r-us/anbox-playstore-installer/

adb: failed to install /home/kivy/Project/iNPPK/mobile/bin/iNPPK-0.3.5-armeabi-v7a-debug.apk: 
Failure [INSTALL_FAILED_OLDER_SDK: Failed parse during installPackageLI: /data/app/vmdl1896004959.tmp/base.apk (at Binary XML file line #11): 
Requires newer sdk version #27 (current version is #25)]

adb uninstall /home/kivy/Project/iNPPK/mobile/bin/iNPPK-0.3.5-armeabi-v7a-debug.apk

под #25 не собирается

клон - kivy-vm-28Oct21-vscode


adb install /home/kivy/Projects/iNPPK/mobile/bin/iNPPK-0.3.5-armeabi-v7a-debug.apk
buildozer android logcat

adb uninstall /home/kivy/Projects/iNPPK/mobile/bin/iNPPK-0.3.5-armeabi-v7a-debug.apk


































---
Из Linux
https://xakep.ru/2017/04/18/anbox/
https://losst.ru/zapusk-android-prilozhenij-na-linux

adb kill-server

sudo adb install /home/kivy/Project/iNPPK/mobile/bin/iNPPK-0.3.5-armeabi-v7a-debug.apk

adb install -r /home/kivy/Project/iNPPK/mobile/bin/iNPPK-0.3.5-armeabi-v7a-debug.apk


После сбоя vbox папка Project недоступна
----
cd /home/kivy/Projects/












--------- 25Oct21
mobile: v.0.3.4 [build 0036]
проверка наличия данных для авторизации

mobile: v.0.3.4 [build 0037]
проверка регистрации на сайте

mobile: v.0.3.5 [build 0038]
переход по ссылке

mobile: v.0.3.5 [build 0039]
---
исправлена ошибка от [build 0029] - работа с прочитанными упрощена, по обновлению меняем статусы просто у всех, а не передаем список просмотренных 
и меняем только у них

mobile: v.0.3.5 [build 0040]
удаление фотографии


--------- 23Oct21
mobile: v.0.3.4 [build 0034] - дизайн от Лизы
mobile: v.0.3.4 [build 0035] - фикс "Warning, too much iteration done before the next frame"


--------- 22Oct21
mobile: v.0.3.4 [build 0033]
dish полностью
php_app: v.0.2.8 [build 0030]
dish полностью

--------- 21Oct21
mobile: v.0.3.3 [build 0032]
actions полностью
backend actions делаем на PHP, т.к. нет работы с SQLite, и нет ДС/МЛ
php_app: v.0.2.7 [build 0029]
actions полностью

--------- 19Oct21
mobile: v.0.3.2 [build 0030]
студенты: доработки интерфейсов
mobile: v.0.3.2 [build 0031]
убрана ошибка "Clock...

--------- 18Oct21
mobile: v.0.3.1 [build 0028]
окончательный вариант показа вакансий
php_app: v.0.2.6 [build 0028]
структура code / action

mobile: v.0.3.2 [build 0029]
read аналогичен вакансиям
----
иногда показываются просмотренные ссылки
php_app: v.0.2.7 [build 0029]
like/dislike read



--------- 16Oct21
mobile: v.0.3.1 [build 0027]
неверная логика: show_vakancy используется и для определения показанных вакансий (для их непоказа) и для
подстчета статистики слов. Надо возвращаться к структуре code / action


--------- 15Oct21
mobile: v.0.3.0 [build 0026]
при обновлении весов слов из ВСЕХ показанных вакансий, то обучение происходит медленно и неточно:
лучше всего обновляются веса распространенных слов
-> поэтому добавляем кнопки изменения весов для КАЖДОЙ вакансии

вывод заголовка вакансии в правильной кодировке в wsgi и в терминале мобильного


--------- 12Oct21
mobile: v.0.2.4 [build 0025]
кнопка "Результаты" в каждой activity


--------- 11Oct21
php_app: v.0.2.5 [build 0027]
для каждой активности свой подход к обновлению весов
show_vacancies - обновляем веса последних показанных


--------- 09Oct21
php_app: v.0.2.4 [build 0024]
поле hour в БД логирования

mobile: v.0.2.4 [build 0024]
доработки интерфейса под концепцию №2

python_app: v.0.2.4 [build 0024]
добавление показанных вакансий в лог (data_1=>vacancy_id)

php_app: v.0.2.4 [build 0025]
неправильное определение wp_id при наличии пустого vk_id

php_app: v.0.2.4 [build 0026]
правильная временная зона



--------- 06Oct21
v. 0.1.2 [build 0021]

mobile/ui/read/read_list.py
import os, requests, json, re, traceback
from kivy.uix.screenmanager import Screen
from kivy.properties import ObjectProperty
import webbrowser




titles_id = {}
items_hrefs = {}


class ReadList(Screen):
    _app = ObjectProperty()
    
    def reload(self):
        db_request = {}
        db_request['tm_id'] = self._app.user_data["tm_id"]
        db_request['vk_id'] = self._app.user_data["vk_id"]

        headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.9; rv:45.0) Gecko/20100101 Firefox/45.0'}
        db_json = json.dumps(db_request)
        
        #отдельный бекенд для вакансий, т.к. он на Python для поиска подходящих вакансий
        page_url = "https://studs.steelfeet.ru/_hack/2020-21/world-it-planet/847-hh-scrapper/next_read.wsgi?q=" + str(db_json)
        t = requests.get(page_url, headers = headers)
        try:
            items_list = json.loads(t.text)
            exist_vacancies_query = items_list["exist_vacancies_query"]
            print()
            print("exist_vacancies_query")
            print(exist_vacancies_query)
        except:
            print()

        self.on_enter()    
    
    def on_enter(self):
        try:
            if (len(self._app.user_data["tm_id"]) < 1) and (len(self._app.user_data["vk_id"]) < 1):
                self._app.screen_manager.current = 'settings_list'
            else:
                db_request = {}
                db_request['code'] = 'read'
                db_request['action'] = 'show'
                db_request['tm_id'] = self._app.user_data["tm_id"]
                db_request['vk_id'] = self._app.user_data["vk_id"]
                headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.9; rv:45.0) Gecko/20100101 Firefox/45.0'}
                db_json = json.dumps(db_request)
                
                page_url = "https://studs.steelfeet.ru/_hack/2020-21/world-it-planet/847-hh-scrapper/best_read.wsgi?q=" + str(db_json)
                t = requests.get(page_url, headers = headers)

                try:
                    items_list = json.loads(t.text)
                    links = items_list["reads"]
                    print()
                    print(items_list["words"])

                except:
                    print(t.text)
                
                #разбиваем title на две строчки
                n = 0
                good_titles = []   
                for item in links:
                    words = item["title"].split(" ")


                    good_title = ""
                    next_br = True
                    for word in words:
                        good_title = good_title + word + " "
                        if len(good_title) > 50:
                            break
                        if (len(good_title) > 25):
                            if (next_br):
                                good_title = good_title + "\n"
                                next_br = False
                                


                                
                                
                    
                    good_titles.append(good_title)
                    titles_id[n] = item["id"]
                    items_hrefs[n] = item["href"]
                    n = n + 1
                
                #выводим на активити
                self.ids.title_0.text = good_titles[0]
                #self.ids.href_0.text = links[0]["href"]

                self.ids.title_1.text = good_titles[1]
                #self.ids.href_1.text = links[1]["href"]

                self.ids.title_2.text = good_titles[2]
                #self.ids.href_2.text = links[2]["href"]

                self.ids.title_3.text = good_titles[3]
                #self.ids.href_3.text = links[3]["href"]

                self.ids.title_4.text = good_titles[4]
                #self.ids.href_4.text = links[4]["href"]















                page_url = "https://studs.steelfeet.ru/_hack/2020-21/world-it-planet/847-hh-scrapper/last_read.wsgi?q=" + str(db_json)
                t = requests.get(page_url, headers = headers)

                items_list = json.loads(t.text)
                links = items_list["reads"]
                
                
                #разбиваем title на две строчки
                for item in links:
                    words = item["title"].split(" ")

                    good_title = ""
                    next_br = True
                    for word in words:
                        good_title = good_title + word + " "
                        if len(good_title) > 50:
                            break
                        if (len(good_title) > 25):
                            if (next_br):
                                good_title = good_title + "\n"
                                next_br = False
                                
                    
                    good_titles.append(good_title)
                    titles_id[n] = item["id"]
                    items_hrefs[n] = item["href"]
                    n = n + 1
                
                #выводим на активити
                self.ids.title_5.text = good_titles[5]
                #self.ids.href_0.text = links[0]["href"]

                self.ids.title_6.text = good_titles[6]
                #self.ids.href_1.text = links[1]["href"]

                self.ids.title_7.text = good_titles[7]
                #self.ids.href_2.text = links[2]["href"]

                self.ids.title_8.text = good_titles[8]
                #self.ids.href_3.text = links[3]["href"]

                self.ids.title_9.text = good_titles[9]
                #self.ids.href_4.text = links[4]["href"]

        except:
            self._app.screen_manager.current = 'settings_list'









    def like_read(self, title_n):
        global titles_id
        try:
            print("read id:" + str(titles_id[title_n]))
            db_request = {}
            db_request['action'] = 'like_read'
            db_request['tm_id'] = self._app.user_data["tm_id"]
            db_request['vk_id'] = self._app.user_data["vk_id"]
            db_request['data_1'] = int(titles_id[title_n])
            db_request['data'] = "data_1=>read_id;"
            headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.9; rv:45.0) Gecko/20100101 Firefox/45.0'}
            db_json = json.dumps(db_request, ensure_ascii=False)
            page_url = "https://steelfeet.ru/app/get.php?q=" + db_json
            t = requests.get(page_url, headers = headers)

        except:
            print("titles_id:")
            print(titles_id)
        

    def dislike_read(self, title_n):
        global titles_id
        print("read id:" + str(titles_id[title_n]))
        
        db_request = {}
        db_request['action'] = 'dislike_read'
        db_request['tm_id'] = self._app.user_data["tm_id"]
        db_request['vk_id'] = self._app.user_data["vk_id"]
        db_request['data_1'] = int(titles_id[title_n])
        db_request['data'] = "data_1=>read_id;"
        headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.9; rv:45.0) Gecko/20100101 Firefox/45.0'}
        db_json = json.dumps(db_request, ensure_ascii=False)
        page_url = "https://steelfeet.ru/app/get.php?q=" + db_json
        t = requests.get(page_url, headers = headers)



    def go_href(self, title_n):
        global items_hrefs
        print("read_href:" + str(items_hrefs[title_n]))
        webbrowser.open_new(str(items_hrefs[title_n]))
